core/src/apps/solana/ui/__init__.py

```diff
@@ -0,0 +1,281 @@
+
+
+def _format_property(value: Any, type: str) -> str | bytes | None:
+    if type in ("pubkey", "authority"):
+        return base58.encode(value)
+    elif type == "lamports":
+        return f"{value / 10 ** 9} SOL"
+    elif type == "unix_timestamp":
+        from trezor.strings import format_timestamp
+
+        return format_timestamp(value)
+    elif isinstance(value, int):
+        return str(value)
+
+    return value
+
+
+def _format_path(path: list[int]) -> str:
+    from apps.common.paths import HARDENED
+
+    if len(path) < 4:
+        return address_n_to_str(path)
+
+    ACCOUNT_PATH_INDEX = 3
+    account_index = path[ACCOUNT_PATH_INDEX]
+    unhardened_account_index: int = account_index ^ (account_index & HARDENED)
+    return f"#{unhardened_account_index + 1}"
+
+
+async def show_confirm(
+    instruction: Instruction,
+    instructions_count: int,
+    instruction_index: int,
+    signer_path: list[int],
+    signer_public_key: bytes,
+) -> None:
+    from trezor.enums import ButtonRequestType
+    from trezor.ui.layouts import confirm_metadata, confirm_properties
+
+    # assertions for pyright
+    assert instruction.parsed_data is not None
+    assert instruction.parsed_accounts is not None
+    assert instruction.ui_properties is not None
+
+    instruction_title = (
+        f"{instruction_index}/{instructions_count}: {instruction.ui_name}"
+    )
+
+    if instruction.is_deprecated_warning is not None:
+        await confirm_metadata(
+            "confirm_deprecated_warning",
+            instruction_title,
+            instruction.is_deprecated_warning,
+            br_code=ButtonRequestType.Other,
+        )
+
+    for ui_property in instruction.ui_properties:
+        if ui_property.parameter is not None:
+            property_template = instruction.get_property_template(ui_property.parameter)
+            value = instruction.parsed_data[ui_property.parameter]
+            _type = property_template.type
+
+            if _type == "authority":
+                if signer_public_key == value:
+                    continue
+
+            await confirm_properties(
+                "confirm_instruction",
+                f"{instruction_index}/{instructions_count}: {instruction.ui_name}",
+                [
+                    (
+                        ui_property.display_name,
+                        _format_property(value, _type),
+                    )
+                ],
+            )
+        elif ui_property.account is not None:
+            account_template = instruction.get_account_template(ui_property.account)
+
+            # optional account, skip if not present
+            if ui_property.account not in instruction.parsed_accounts:
+                continue
+
+            account_value = instruction.parsed_accounts[ui_property.account]
+
+            if account_template.is_authority:
+                if signer_public_key == account_value[0]:
+                    continue
+
+            account_data = []
```

Comment:
Could be type-hinted, just to be more safe

Reply:
[Changes](https://github.com/trezor/trezor-firmware/commit/236b05ecea4131347737d54350c5408d7a082b19)